### PR TITLE
Feat/more metrics

### PR DESCRIPTION
**Pull request by wileyj**
## Description

This PR adds more metrics to a stacks-node, enabling easier instrumentation around what's happening on the node. 
This was necessary because it's not always feasible or practical to request the data directly from the node's RPC endpoints or elsewhere. 
This builds upon previous work done by @lgalabru

For details refer to PR https://github.com/blockstack/stacks-blockchain/pull/1670

Sample of metrics output:
```
# HELP stacks_node_active_miners_total Total number of active miners
# TYPE stacks_node_active_miners_total gauge
stacks_node_active_miners_total 7
# HELP stacks_node_bandwidth_inbound Total inbound bandwidth total in bytes
# TYPE stacks_node_bandwidth_inbound gauge
stacks_node_bandwidth_inbound 6410139
# HELP stacks_node_bandwidth_outbound Total outbound bandwidth total in bytes
# TYPE stacks_node_bandwidth_outbound gauge
stacks_node_bandwidth_outbound 2404165
# HELP stacks_node_btc_blocks_received_total Total number of blocks processed from the burnchain
# TYPE stacks_node_btc_blocks_received_total counter
stacks_node_btc_blocks_received_total 366
# HELP stacks_node_burn_block_height Burnchain tip height
# TYPE stacks_node_burn_block_height gauge
stacks_node_burn_block_height 673891
# HELP stacks_node_mempool_gc_count Total count of all mempool garbage collections
# TYPE stacks_node_mempool_gc_count counter
stacks_node_mempool_gc_count 489
# HELP stacks_node_mempool_outstanding_txs Number of still-unprocessed transactions received by this node since it started
# TYPE stacks_node_mempool_outstanding_txs gauge
stacks_node_mempool_outstanding_txs{handler="all"} 16
# HELP stacks_node_mempool_tx_confirm_times Time (seconds) between when a tx was received by this node's mempool and when a tx was first processed in a block
# TYPE stacks_node_mempool_tx_confirm_times histogram
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="300"} 5
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="600"} 6
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="900"} 23
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="1200"} 44
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="1500"} 54
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="1800"} 77
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="2100"} 88
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="2400"} 88
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="2700"} 92
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="3000"} 94
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="3600"} 106
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="4200"} 108
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="4800"} 108
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="6000"} 108
stacks_node_mempool_tx_confirm_times_bucket{handler="all",le="+Inf"} 108
stacks_node_mempool_tx_confirm_times_sum{handler="all"} 172282
stacks_node_mempool_tx_confirm_times_count{handler="all"} 108
# HELP stacks_node_message_count Stacks message count by type of message
# TYPE stacks_node_message_count counter
stacks_node_message_count{name="p2p_authenticated_handshake"} 5
stacks_node_message_count{name="p2p_get_neighbors"} 5
# HELP stacks_node_neighbors_inbound Total count of current known inbound neighbors
# TYPE stacks_node_neighbors_inbound gauge
stacks_node_neighbors_inbound 0
# HELP stacks_node_neighbors_outbound Total count of current known outbound neighbors
# TYPE stacks_node_neighbors_outbound gauge
stacks_node_neighbors_outbound 19
# HELP stacks_node_rpc_bandwidth_inbound Total RPC inbound bandwidth in bytes
# TYPE stacks_node_rpc_bandwidth_inbound gauge
stacks_node_rpc_bandwidth_inbound 37389
# HELP stacks_node_rpc_request Total number of RPC requests made by path
# TYPE stacks_node_rpc_request counter
stacks_node_rpc_request{method="get",path="/v2/info"} 3
# HELP stacks_node_rpc_request_total Total number of RPC requests made
# TYPE stacks_node_rpc_request_total counter
stacks_node_rpc_request_total 3
# HELP stacks_node_smart_contracts Total count of smart contracts
# TYPE stacks_node_smart_contracts counter
stacks_node_smart_contracts 11
# HELP stacks_node_stacks_tip_height Stacks chain tip height
# TYPE stacks_node_stacks_tip_height gauge
stacks_node_stacks_tip_height 6507
# HELP stacks_node_stx_blocks_processed_total Total number of stacks blocks processed
# TYPE stacks_node_stx_blocks_processed_total counter
stacks_node_stx_blocks_processed_total 10
# HELP stacks_node_stx_blocks_received_total Total number of Stacks blocks received
# TYPE stacks_node_stx_blocks_received_total counter
stacks_node_stx_blocks_received_total 1
# HELP stacks_node_transactions_received_total Total number of transactions received and relayed
# TYPE stacks_node_transactions_received_total counter
stacks_node_transactions_received_total 572
```
## Type of Change
- [X] New feature
- [ ] Bug fix
- [ ] API reference/documentation update
- [X] Other

### CODE REVIEW

File: src/burnchains/db.rs
Comment (jcnelson) [inline src/burnchains/db.rs:154]: I don't think this is the right place for this.  I think this should happen in the top-level `Burnchain::sync()` method.  (Really, I think this should happen in the chains coordinator, but the chains coordinator doesn't drive burnchain synchronization yet).

File: src/chainstate/burn/operations/stack_stx.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/stack_stx.rs:323]: This should be somewhere in the top-level `Burnchain::sync()` method, such as right after when a block gets processed.  There's no guarantee anywhere that this method will be called only once.

File: src/chainstate/burn/operations/transfer_stx.rs
Comment (jcnelson) [inline src/chainstate/burn/operations/transfer_stx.rs:229]: This should be somewhere in the top-level `Burnchain::sync()` method, such as right after when a block gets processed.  There's no guarantee anywhere that this method will be called only once.

File: src/chainstate/stacks/address.rs
Comment (jcnelson) [inline src/chainstate/stacks/address.rs:210]: The method `to_b58()` is used in debugging and would get called all over the place, so this is probably isn't a good place for this either.  What are you trying to track here?

File: src/chainstate/stacks/transaction.rs
Comment (jcnelson) [inline src/chainstate/stacks/transaction.rs:253]: This method I don't think is used anywhere besides the test code and client-side tooling.  Are you trying to count the number of smart contracts?  If so, then what you should probably do is look at the event receipts emitted in `process_ready_blocks()` in the chains coordinator in order to see which smart contracts get instantiated (and just count those events).

File: src/core/mempool.rs
Comment (jcnelson) [inline src/core/mempool.rs:561]: Are you trying to get the total size of the mempool here?  If so, then this doesn't do what you think it does.  The `get_txs_at()` method only returns the number of transactions that arrived while the given `tip_consensus_hash` and `tip_block_hash` were the chain tip.  I think you'll want to add a method that returns the total number of rows in the mempool, and then call this `update_stx_mempool_size()` method within the relayer (right after it does a mempool garbage-collection).

File: src/monitoring/mod.rs
Comment (jcnelson) [inline src/monitoring/mod.rs:20]: Are we going to add this commented code back at some point?  If not, then let's just delete it (same below).

File: src/net/p2p.rs
Comment (jcnelson) [inline src/net/p2p.rs:918]: This probably isn't what you want.  The `outbound_neighbors` and `inbound_neighbors` here subsets of the total inbound/outbound set sizes, since nodes don't broadcast messages to all peers.  I think you'll want to move these counters to the top-level P2P runloop, right after the call to `prune_connections()`.  You'd use the `count_outbound_conversations()` method to count the number of outbound neighbors, and then take `self.peers.len() - self.count_outbound_conversations(..)` to get the number of inbound neighbors.

File: src/burnchains/db.rs
Comment (wileyj) [reply src/burnchains/db.rs]: This has been moved to `src/burnchains/burnchain.rs`

File: src/chainstate/stacks/transaction.rs
Comment (wileyj) [reply src/chainstate/stacks/transaction.rs]: Moved to `src/chainstate/coordinator/mod.rs`

File: src/core/mempool.rs
Comment (wileyj) [reply src/core/mempool.rs]: Correct, the idea here was to track the current size of the mempool. I've removed this for now, will look into this more at a later time.

File: src/monitoring/mod.rs
Comment (wileyj) [reply src/monitoring/mod.rs]: Comments have been removed

File: src/net/p2p.rs
Comment (wileyj) [reply src/net/p2p.rs]: Moved: https://github.com/wileyj/stacks-blockchain/commit/192b66fe0bd669fb7bf3177034256d5383c4bdaa

Also moved `update_burnchain_height`, since it wasn't  reliably being updated on restart. Seems to work better now and burn_height is being reported reliably after a restart.

File: src/chainstate/burn/operations/transfer_stx.rs
Comment (wileyj) [reply src/chainstate/burn/operations/transfer_stx.rs]: Removed

**Comment by wileyj** [comment]
Merged with latest develop branch which has some additional metrics added since this PR was was sent

**Comment by wileyj** [comment]
@jcnelson @kantai

File: src/chainstate/coordinator/mod.rs
Comment (kantai) [inline src/chainstate/coordinator/mod.rs:691]: This metric is more accurately "contract calls processed" , rather than "stx_smart_contracts".

File: src/net/rpc.rs
Comment (kantai) [inline src/net/rpc.rs:115]: Can you delete the commented out lines?

**Review by kantai — COMMENTED** [review]
These changes look good to me -- my only comments are pretty light:
* clean up the commented out code segments
* revert the changes in `net-test`
* the `stx_smart_contracts` counter isn't counting what it's described as: it should be labeled as "contract calls processed".

**Review by jcnelson — APPROVED** [review]
LGTM!  Just be sure to address @kantai's feedback before merging.  Thanks!

File: src/chainstate/coordinator/mod.rs
Comment (wileyj) [reply src/chainstate/coordinator/mod.rs:691]: renamed to `increment_contract_calls_processed`

File: src/net/rpc.rs
Comment (wileyj) [reply src/net/rpc.rs:115]: removed

**Comment by wileyj** [comment]
> These changes look good to me -- my only comments are pretty light:
> 
> * clean up the commented out code segments
> * revert the changes in `net-test`
> * the `stx_smart_contracts` counter isn't counting what it's described as: it should be labeled as "contract calls processed".

Reverted the `net-test` change - this was from some comments around using testground more effectively, but can be addressed in a different PR.

**Comment by wileyj** [comment]
@kantai @lgalabru

File: src/monitoring/prometheus.rs
Comment (lgalabru) [inline src/monitoring/prometheus.rs:92]: In theory, this number can only go up - should it be a gauge?

File: src/monitoring/prometheus.rs
Comment (lgalabru) [inline src/monitoring/prometheus.rs:97]: In theory, this number can only go up - should it be a gauge?

File: src/monitoring/prometheus.rs
Comment (lgalabru) [reply src/monitoring/prometheus.rs:92]: Nvm, going with a gauge is probably more convenient.

File: src/monitoring/prometheus.rs
Comment (lgalabru) [reply src/monitoring/prometheus.rs:97]: Nvm, going with a gauge is probably more convenient.

File: src/net/rpc.rs
Comment (lgalabru) [inline src/net/rpc.rs:618]: I don't know what's the right way to resolve this, but we've been implementing the same thing, with 2 different approaches - https://github.com/blockstack/stacks-blockchain/pull/2453/commits/134671da798a61cf513672711c0f2ea8fb24fe7a.

**Review by lgalabru — APPROVED** [review]
Looks good to me, thank you @wileyj!

**Comment by wileyj** [comment]
> Looks good to me, thank you @wileyj!

Will defer to upcoming PR @lgalabru for these changes

**Review by kantai — APPROVED** [review]

**Pull request merged**
